### PR TITLE
Initialize Selections

### LINKED ISSUE #3409: Selecting values upon start in a barchart
Hi, 

I would like to pre-select certain values in a barchart up loading.

How would I do that?

For Example:

Selecting in advance Feb 2018 in this graph

image

Current spec:

var vlSpec =
        {
            "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
            "data": { "url": "php/solar.php" },
            "vconcat": [{
                "width": 1000,
                "mark": "area",
                "encoding": {
                    "x": {
                        "field": "date",
                        "type": "temporal",
                        "scale": { "domain": { "selection": "brush" } },
                        "axis": { "title": "" }
                    },
                    "y": { "field": "price", "type": "quantitative" }
                }
            }, {
                "width": 1000,
                "height": 50,
                "mark": "area",
                "selection": {
                    "brush": { "type": "interval", "encodings": ["x"] }
                },
                "encoding": {
                    "x": {
                        "field": "date",
                        "type": "temporal",
                        "axis": { "format": "%Y-%m" }
                    },
                    "y": {
                        "field": "price",
                        "type": "quantitative",
                        "axis": { "tickCount": 3, "grid": false }
                    }
                }
            }]
        }

### PR DESCRIPTION
**Pull request by arvind**
**This PR relies on the refactoring introduced by #4068, so please merge that one first.**

This PR adds an `init` property to selection definitions to initialize them using the fields or encoding channels a selection is projected over. For example, here is an updated version of our `interactive_query_widgets` example that initializes the selection to select cars with 8 cylinders manufactured in 1971. 

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "description": "Drag the sliders to highlight points.",
  "data": {"url": "data/cars.json"},
  "transform": [{"calculate": "year(datum.Year)", "as": "Year"}],
  "layer": [{
    "selection": {
      "CylYr": {
        "type": "single", "fields": ["Cylinders", "Year"],
        "init": {"Cylinders": 4, "Year": 1977},
        "bind": {
          "Cylinders": {"input": "range", "min": 3, "max": 8, "step": 1},
          "Year": {"input": "range", "min": 1969, "max": 1981, "step": 1}
        }
      }
    },
    "mark": "circle",
    "encoding": {
      "x": {"field": "Horsepower", "type": "quantitative"},
      "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
      "color": {
        "condition": {"selection": "CylYr", "field": "Origin", "type": "nominal"},
        "value": "grey"
      }
    }
  }, {
    "transform": [{"filter": {"selection": "CylYr"}}],
    "mark": "circle",
    "encoding": {
      "x": {"field": "Horsepower", "type": "quantitative"},
      "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
      "color": {"field": "Origin", "type": "nominal"},
      "size": {"value": 100}
    }
  }]
}
```

Similarly, an updated `interactive_brush` example to initialize the brush selection is as follows:

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "description": "Drag out a rectangular brush to highlight points.",
  "data": {"url": "data/cars.json"},
  "selection": {
    "brush": {
      "type": "interval",
      "init": {"x": [55, 160], "y": [13, 37]}
    }
  },
  "mark": "point",
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {
      "condition": {"selection": "brush", "field": "Cylinders", "type": "ordinal"},
      "value": "grey"
    }
  }
}
```

Fixes #3409, #3431, and altair-viz/altair#599. 

/cc @jakevdp

### CODE REVIEW

**Comment by arvind** [comment]
Tagged as blocked until #4068 is merged (which itself is blocked by vega/vega-parser#89). But, once the upstream PRs are merged, this one should be good to go. Compile- and runtime tests are passing locally.

**Comment by arvind** [comment]
After investigating this issue, I suspect there may be a bug in Vega (vega/vega#1428) where signals continue to listen for scale updates even if they're marked as `react: false`.

**Comment by arvind** [comment]
Hooray, with Vega 4.3, this PR is no longer blocked and ready for review as well @domoritz, @kanitw!

**Comment by arvind** [comment]
@domoritz: As there'll be several documentation updates related to all the new selections work, I wanted to do all of it in a separate more comphreneisve pass/PR rather than piecemeal if that's ok.

**Comment by domoritz** [comment]
That makes sense. If you think this will take more than a week, please create an issue that describes what needs to be documented.

File: examples/specs/interactive_brush.vl.json [hunk 1]
```diff
@@ -4,7 +4,8 @@
   "data": {"url": "data/cars.json"},
   "selection": {
     "brush": {
-      "type": "interval"
+      "type": "interval",
+      "init": {"x": [55, 160], "y": [13, 37]}
```
Comment (domoritz) [inline examples/specs/interactive_brush.vl.json]: Shouldn't we see the brush in the SVG as well? See https://github.com/vega/vega-lite/pull/4139/files?short_path=b21ff41#diff-b21ff41954cf78b8b1141416d74085d5

**Comment by domoritz** [comment]
Let me know when you get the chance to update the PR. I'd like to make a release soon so we can test these changes thoroughly.

**Comment by arvind** [comment]
Supporting datetime objects proved trickier than expected since it involves juggling expression strings with regular JSON primitives. But, I think I got it working. And also extended initialization to support initializing multi selections with an array of values. 

Unfortunately, merging this issue is blocked by three issues/PRs in Vega: 
1. vega/vega#1433 which produces a bug where the brush mark doesn't appear for initialized interval selections (introduced, ironically, when vega/vega#1428 was fixed)
2. vega/vega#1435 which produces a bug when initializing a bound selection
3. vega/vega-util#10 which provides correct typings for the `identity` function, and without which the CI will not run succesfully.

**Comment by domoritz** [comment]
Okay, let's wait for a Vega release then. Luckily, no other PR is blocked by this and we can make a new Vega-Lite rc release without this PR.

**Comment by eharkins** [comment]
To clarify, the solution involving the `init` property, that will solve problems like https://github.com/vega/vega/issues/1435, will also be added to Vega in addition to Vega-Lite once you have a chance to merge this and/or a Vega version of the same PR? Looking forward to having this functionality in Vega! Thanks

**Comment by arvind** [comment]
Yes, that's right @eharkins!

**Comment by domoritz** [comment]
What feature is this PR blocked by right now?

**Comment by arvind** [comment]
vega/vega-util#10.

**Comment by domoritz** [comment]
Ping @jheer about https://github.com/vega/vega-util/pull/10. Luckily, I think we should be able to update vega-util without a new Vega release.

File: src/compile/selection/selection.ts [hunk 2]
```diff
@@ -410,3 +415,11 @@ export function positionalProjections(selCmpt: SelectionComponent) {
   });
   return {x, xi, y, yi};
 }
+
+export function assembleInit(init: any, wrap: (str: string) => string = identity): string {
```
Comment (domoritz) [inline src/compile/selection/selection.ts]: Why is `init` `any` here? Can we use a stricter type?

File: src/selection.ts [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
 import {VgEventStream} from './vega.schema';
 
 export const SELECTION_ID = '_vgsid_';
 export type SelectionType = 'single' | 'multi' | 'interval';
 export type SelectionResolution = 'global' | 'union' | 'intersect';
 
+export type SelectionInit = boolean | number | string | DateTime;
+export type SelectionInitArray = boolean[] | number[] | string[] | DateTime[];
+export interface SelectionInitMapping {
+  [key: string]: SelectionInit;
```
Comment (domoritz) [inline src/selection.ts]: is the key here a selection name? If so, let's add a comment.

Comment (domoritz) [reply src/selection.ts]: or a field name?

File: test/compile/selection/inputs.test.ts [hunk 4]
```diff
@@ -154,4 +165,32 @@ describe('Inputs Selection Transform', () => {
       ])
     );
   });
+
+  it('respects initialization', () => {
+    model.component.selection = {seven: selCmpts['seven']};
+    expect(selection.assembleUnitSelectionSignals(model, [])).toEqual(
```
Comment (domoritz) [inline test/compile/selection/inputs.test.ts]: You can just use `expect.toContainEqual(...)` here

File: src/compile/selection/selection.ts (hunk 2 above)
Comment (arvind) [reply src/compile/selection/selection.ts]: It should be `SelectionInit | SelectionInitArray` but I'm not able to figure out why TypeScript throws a big error with the `map` call on the next line... `isArray` sets up a type guard...

Comment (domoritz) [reply src/compile/selection/selection.ts]: Fixed in 7f796587b

File: build/vega-lite-schema.json [hunk 5]
```diff
@@ -5146,6 +5146,13 @@
           },
           "type": "array"
         },
+        "init": {
+          "additionalProperties": {
+            "$ref": "#/definitions/SelectionInitArray"
+          },
+          "description": "Initialize the selection with a mapping between field names and arrays of\ninitial values.",
```
Comment (kanitw) [inline build/vega-lite-schema.json]: From your examples in the main conversation page, it seems like `init` supports both (a) a mapping between field names and arrays of initial values and (b) a mapping between channels and arrays of initial values?

If so, how does `init` know if a key is a channel or field name?
(For example, what if my `data` is `[{x:1, y:2}, ...]` -- how do I know if `"x"` is a channel or a field?

**Comment by domoritz** [comment]
Waiting for https://github.com/vega/vega-lite/pull/4385

**Comment by domoritz** [comment]
Waiting for docs before we merge this.

File: src/compile/selection/selection.ts [hunk 6]
```diff
@@ -413,3 +418,14 @@ export function positionalProjections(selCmpt: SelectionComponent) {
   });
   return {x, xi, y, yi};
 }
+
+export function assembleInit(
+  init: SelectionInit | SelectionInitArray,
+  wrap: (str: string) => string = identity
+): string {
+  return isArray(init)
+    ? `[${(init as SelectionInit[]).map(v => assembleInit(v, wrap)).join(', ')}]`
```
Comment (kanitw) [inline src/compile/selection/selection.ts]: @arvind I've try to resolve conflicts in this branch with master. 
However, I'm getting some type errors, and it's unclear to me why we should to the casting (`as SelectionInit[]`) on this line.  

Shouldn't we fix the type of the `init` above?

File: examples/compiled/interactive_panzoom_splom.vg.json [hunk 7]
```diff
@@ -273,13 +275,20 @@
                 {"signal": "grid_Acceleration"},
                 {"signal": "grid_Horsepower"}
               ],
+<<<<<<< HEAD
```
Comment (kanitw) [inline examples/compiled/interactive_panzoom_splom.vg.json]: Note that I can't build examples yet as I still build errors due to typing.

File: src/selection.ts [hunk 8]
```diff
@@ -163,6 +181,12 @@ export interface IntervalSelectionConfig extends BaseSelectionDef {
    * appearance of the mark.
    */
   mark?: BrushConfig;
+
+  /**
+   * Initialize the selection with a mapping between field names and arrays of
+   * initial values.
+   */
+  init?: {[key: string]: SelectionInitArray};
```
Comment (kanitw) [inline src/selection.ts]: We have 

```
export interface SelectionInitMapping {
  [key: string]: SelectionInit;
}
```

Should this be define as `SelectionInitArrayMapping` too, for consistency?

File: src/selection.ts [hunk 9]
```diff
@@ -63,6 +70,11 @@ export interface SingleSelectionConfig extends BaseSelectionDef {
    * See the [nearest transform](https://vega.github.io/vega-lite/docs/nearest.html) documentation for more information.
    */
   nearest?: boolean;
+
+  /**
+   * Initialize the selection with a mapping between field names and initial values.
```
Comment (kanitw) [inline src/selection.ts]: > field names?

(Or channels?)

File: src/selection.ts [hunk 10]
```diff
@@ -85,6 +97,12 @@ export interface MultiSelectionConfig extends BaseSelectionDef {
    * See the [nearest transform](https://vega.github.io/vega-lite/docs/nearest.html) documentation for more information.
    */
   nearest?: boolean;
+
+  /**
+   * Initialize the selection with a mapping between field names and an initial
```
Comment (kanitw) [inline src/selection.ts]: > field names?

(Or channels?)

File: src/selection.ts (hunk 8 above)
Comment (kanitw) [reply src/selection.ts]: > field names?

(Or channels?)

File: src/compile/selection/selection.ts (hunk 6 above)
Comment (kanitw) [reply src/compile/selection/selection.ts]: Actually, I just realize the Dom is the one who wrote this line.

File: src/selection.ts (hunk 10 above)
Comment (domoritz) [reply src/selection.ts]: Shouldn't channels have higher precedence?

File: src/compile/selection/selection.ts (hunk 6 above)
Comment (kanitw) [reply src/compile/selection/selection.ts]: Ok fixed.

File: src/compile/selection/transforms/project.ts [hunk 11]
```diff
@@ -72,12 +73,23 @@ const project: TransformCompiler = {
...
+        }
+
+        if (isIntervalSelection(selDef)) {
+          const initCmpt = parseInit(selDef.init);
+          selCmpt.init = initCmpt;
+        } else {
+          const init = isArray(selDef.init) ? selDef.init : [selDef.init];
+          const initCmpt = init.map(parseInit);
+
+          selCmpt.init = initCmpt;
```
Comment (kanitw) [inline src/compile/selection/transforms/project.ts]: @arvind I tried to modify the code on the left to have more strict typings, and it seems like the logic is incorrect as it does not produce a matching type.  

For this `else` case, the initCmpt is `SelectionInit[][]`, but `init` of a selection component expects `(SelectionInit | SelectionInitArray)[]`.  So I wonder if you forget to flatten here.

Comment (kanitw) [reply src/compile/selection/transforms/project.ts]: NVM, I think I got the logic now. Let me fix and push.

**Comment by kanitw** [comment]
I just resolved conflicts and clean up the typings (while trying to preserve the logic).  

@domoritz and @arvind -- can one of you double check so we can merge this for 3.0?

File: src/compile/selection/transforms/project.ts [hunk 12]
```diff
@@ -72,12 +73,20 @@ const project: TransformCompiler = {
...
-        const parseInit = (i: any) => proj.map(p => (i[p.channel] !== undefined ? i[p.channel] : i[p.field]));
-        selCmpt.init = selCmpt.type === 'interval' ? parseInit(init) : array(init).map(parseInit);
+        function parseInit<T extends SelectionInitMapping | SelectionInitArrayMapping>(i: T): T['a'][] {
+          return proj.map(p => (i[p.channel] !== undefined ? i[p.channel] : i[p.field]));
+        }
+
+        if (isIntervalSelection(selDef)) {
+          selCmpt.init = parseInit(selDef.init);
+        } else {
+          const init = isArray(selDef.init) ? selDef.init : [selDef.init];
```
Comment (kanitw) [inline src/compile/selection/transforms/project.ts]: We could use array again here once we have https://github.com/vega/vega/pull/1618 in.

File: src/selection.ts [hunk 13]
```diff
@@ -1,11 +1,22 @@
...
 export type SelectionResolution = 'global' | 'union' | 'intersect';
 
+export type SelectionInit = boolean | number | string | DateTime;
+export type SelectionInitArray = boolean[] | number[] | string[] | DateTime[];
+export interface SelectionInitMapping {
+  [key: string]: SelectionInit;
+}
+
+export interface SelectionInitArrayMapping {
+  [key: string]: SelectionInitArray;
```
Comment (domoritz) [inline src/selection.ts]: The key here is either a field or a channel, right? Can we make that explicit?

File: test/compile/selection/inputs.test.ts [hunk 14]
```diff
@@ -154,4 +165,32 @@ describe('Inputs Selection Transform', () => {
...
+      expect.arrayContaining([
+        {
+          name: 'seven_tuple',
+          update: 'seven_Year !== null ? {fields: seven_tuple_fields, values: [seven_Year]} : null'
+        }
+      ])
```
Comment (domoritz) [inline test/compile/selection/inputs.test.ts]: is only one entry here. Why use arrayContaining?

**Comment by domoritz** [comment]
Not sure why the yarn lockfile needs to change but okay.

File: src/selection.ts (hunk 13 above)
Comment (kanitw) [reply src/selection.ts]: Sorry it doesn't work, which means we can merge at this point. 

image

File: test/compile/selection/inputs.test.ts (hunk 14 above)
Comment (kanitw) [reply test/compile/selection/inputs.test.ts]: This is not created in this PR, so feel free to send a follow up fix.

File: src/selection.ts (hunk 13 above)
Comment (domoritz) [reply src/selection.ts]: Well, was an idea.

**Comment by kanitw** [comment]
> Not sure why the yarn lockfile needs to change but okay.

Good catch! I think that's a result of auto-conflict merge.  I'll remove it.

File: src/compile/selection/transforms/project.ts (hunk 12 above)
Comment (domoritz) [reply src/compile/selection/transforms/project.ts]: That relies on the Vega 5 transition to let's do it later.

**Comment by kanitw** [comment]
:tada:

File: src/compile/selection/selection.ts [hunk 15]
```diff
@@ -413,3 +425,16 @@ export function positionalProjections(selCmpt: SelectionComponent) {
...
+
+export function assembleInit(
+  init: (SelectionInit | SelectionInit[] | SelectionInitArray)[] | SelectionInit,
+  wrap: (str: string) => string = identity
+): string {
+  if (isArray(init)) {
+    const str = init.map(v => assembleInit(v, wrap)).join(', ');
+    return `[${str}]`;
+  } else if (isDateTime(init)) {
+    return wrap(dateTimeExpr(init));
```
Comment (domoritz) [inline src/compile/selection/selection.ts]: Shouldn't this be `dateTimeExpr(init, true)`?

**Pull request merged**